pinot-broker/src/main/java/org/apache/pinot/broker/routing/RemoteClusterBrokerRoutingManager.java

```diff
@@ -0,0 +1,123 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.apache.pinot.common.metadata.ZKMetadataProvider;
+import org.apache.pinot.common.metrics.BrokerMetrics;
+import org.apache.pinot.common.metrics.BrokerTimer;
+import org.apache.pinot.core.transport.server.routing.stats.ServerRoutingStatsManager;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The {@code RemoteClusterBrokerRoutingManager} manages the routing for a single remote Pinot cluster available for
+ * federation. It periodically checks for changes in the set of tables available in the remote cluster and updates the
+ * routing accordingly.
+ */
+public class RemoteClusterBrokerRoutingManager extends BrokerRoutingManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RemoteClusterBrokerRoutingManager.class);
+  private static final long ROUTING_CHANGE_DETECTION_INTERVAL_MS = 10_000L;
+
+  private final String _remoteClusterName;
+  private final AtomicBoolean _processChangeInRouting = new AtomicBoolean(false);
+  private final ScheduledExecutorService _routingChangeExecutor;
+
+  public RemoteClusterBrokerRoutingManager(String remoteClusterName, BrokerMetrics brokerMetrics,
+      ServerRoutingStatsManager serverRoutingStatsManager, PinotConfiguration pinotConfig) {
+    super(brokerMetrics, serverRoutingStatsManager, pinotConfig);
+    _remoteClusterName = remoteClusterName;
+    _routingChangeExecutor = Executors.newSingleThreadScheduledExecutor();
+    _routingChangeExecutor.scheduleAtFixedRate(this::determineRoutingChangeForTables, 0,
+        ROUTING_CHANGE_DETECTION_INTERVAL_MS, TimeUnit.MILLISECONDS);
+  }
+
+  public void shutdown() {
+    _routingChangeExecutor.shutdownNow();
+  }
+
+  public void determineRoutingChangeForTables() {
```

Comment:
if this throws the error won't be logged I think since this is called from an executor service.

Can you wrap it in try-finally?

```
long startTime = ;
try {
} finally {
  brokerMetrics...
}
```